matchzoo/preprocessors.py

```diff
@@ -1,146 +1,165 @@
 
-    :param tokens: list of tokenized tokens.
-    :param lang: language code for stopwords.
 
-    :return tokens: list of tokenized tokens without stopwords.
-    """
-    return [token for token in tokens if token not in get_stopwords(lang)]
+class PuncRemovalUnit(StatelessProcessUnit):
+    """Process unit for remove punctuations."""
 
+    def transform(self, tokens: list) -> list:
+        """
+        Remove punctuations from list of tokens.
 
-def remove_punctuation(tokens: list) -> list:
-    """
-    Remove punctuations from list of tokens.
+        :param tokens: list of toekns.
 
-    :param tokens: list of
+        :return rv: tokens  without punctuation.
+        """
+        return [token for token in tokens if not match_punc.search(token)]
 
-    :return rv: tokens  without punctuation.
-    """
-    return [token for token in tokens if not match_punc.search(token)]
 
+class DigitRemovalUnit(StatelessProcessUnit):
+    """Process unit to remove digits."""
 
-def remove_digits(tokens: list) -> list:
-    """
-    Remove digits from list of tokens.
+    def transform(self, tokens: list) -> list:
+        """
+        Remove digits from list of tokens.
 
-    :param tokens: list of tokens to be filtered.
+        :param tokens: list of tokens to be filtered.
 
-    :return tokens: tokens of tokens without digits.
-    """
-    return [token for token in tokens if not token.isdigit()]
+        :return tokens: tokens of tokens without digits.
+        """
+        return [token for token in tokens if not token.isdigit()]
 
 
-def stemming(tokens: list, stemmer: str = 'porter') -> list:
-    """
-    Reducing inflected words to their word stem, base or root form.
+class StopRemovalUnit(StatelessProcessUnit):
+    """Process unit to remove stop words."""
 
-    :param tokens: list of string to be stemmed.
+    def transform(self, tokens: list, **kwargs: dict) -> list:
+        """
+        Remove stopwords from list of tokenized tokens.
 
-    :raise ValueError: stemmer type should be porter or lancaster.
+        :param tokens: list of tokenized tokens.
+        :param lang: language code for stopwords.
 
-    :return tokens: stemmed token.
-    """
-    if stemmer == 'porter':
-        porter_stemmer = nltk.stem.PorterStemmer()
-        return [porter_stemmer.stem(token) for token in tokens]
-    elif stemmer == 'lancaster' or stemmer == 'krovetz':
-        lancaster_stemmer = nltk.stem.lancaster.LancasterStemmer()
-        return [lancaster_stemmer.stem(token) for token in tokens]
-    else:
-        raise ValueError(
-            'Not supported supported stemmer type: {}'.format(stemmer))
+        :return tokens: list of tokenized tokens without stopwords.
+        """
+        lang = kwargs.get('lang', 'en')
+        return [token
+                for token
+                in tokens
+                if token not in StopRemovalUnit._get_stopwords(lang)]
 
+    @staticmethod
+    def _get_stopwords(lang: str) -> list:
```

Comment:
If `lang` is set in `__init__`, this can be converted to a normal method that takes no argument, which is usually better than a **private** static method, which you're going to use else where (static methods sometimes are purely for namespaces).

Reply:
okay